cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -529,47 +541,118 @@ public Void call() throws Exception {
     });
   }
 
-  private void setInputDatasetIfNeeded(Job job) throws IOException {
-    InputFormatProvider provider = context.getInputFormatProvider();
+  // Multiple mappers could be defined, so we check that their output types are consistent.
+  // Return the TypeToken for one of the mappers (doesn't matter which one, since we check that all of their output
+  // key/value types are consistent).
+  // Returns null if their are no inputs configured.
+  @Nullable
+  private TypeToken<?> checkMapperOutputTypes(Job job) throws ClassNotFoundException {
+    int numInputs = context.getMapperInputs().size();
+    if (numInputs == 0) {
+      // if there are no inputs configured through CDAP APIs, nothing to check for consistency
+      return null;
+    }
+
+    // get all the mappers configured through our APIs
+    List<Class<? extends Mapper>> mapperClasses = new ArrayList<>();
+    for (MapperInput mapperInput : context.getMapperInputs().values()) {
+      Class<? extends Mapper> mapper = mapperInput.getMapper();
+      if (mapper != null) {
+        mapperClasses.add(mapper);
+      }
+    }
+
+    if (mapperClasses.size() < numInputs) {
+      // means one of the inputs is going to default to use the mapper assigned on the job,
+      // so also check the job's mapper class
+      if (!isProgrammaticConfig(job.getConfiguration(), Job.MAP_CLASS_ATTR)) {
+        // mapper class was not explicitly set on the job (reduce only job)
+        return null;
+      }
+      mapperClasses.add(job.getMapperClass());
+    }
+
+    Set<Class<?>> mapOutputKeys = new HashSet<>();
```

Comment:
You could do this in a way that produces a better error message, one that indicates the two mapper classes and their types that conflict. By maintaining, instead of a set, a single key or value class, and instead of adding to a set, every time check whether it is the same class. I think that would make the error message a lot more meaningful